runtime/interpreter/value.go

```diff
@@ -6487,6 +6544,72 @@ func (v *CompositeValue) ConformsToDynamicType(
 	return true
 }
 
+func (v *CompositeValue) Fields() *StringValueOrderedMap {
+	v.ensureFieldsLoaded()
+	return v.fields
+}
+
+func (v *CompositeValue) Location() common.Location {
+	v.ensureMetaInfoLoaded()
+	return v.location
+}
+
+func (v *CompositeValue) QualifiedIdentifier() string {
+	v.ensureMetaInfoLoaded()
+	return v.qualifiedIdentifier
+}
+
+func (v *CompositeValue) Kind() common.CompositeKind {
+	v.ensureMetaInfoLoaded()
+	return v.kind
+}
+
+// ensureMetaInfoLoaded ensures loading the meta information of this composite value.
+// If the meta info is already loaded, then calling this function won't have any effect.
+// Otherwise, the values are decoded form the cached raw-content.
+//
+// Meta info includes:
+//    - location
+//    - qualifiedIdentifier
+//    - kind
+//
+func (v *CompositeValue) ensureMetaInfoLoaded() {
+	if v.content == nil {
+		return
+	}
+
+	err := decodeCompositeMetaInfo(v, v.content)
+	if err != nil {
+		panic(err)
+	}
+
+	// Raw content is no longer needed. Clear the cache and free-up the memory.
```

Comment:
What do you think about postponing the release of v.content until it is modified?  That way, we don't have to re-encode them if they are stored without changes.

Reply:
That's a good idea. However, the catch is, fields also could get indirectly modified. e.g: Updating a field of a nested composite value. For the outer value, it would only see it as a field-retrieval. i.e: `foo.bar.someField = "updated"`
Maybe if we had something like a 'update callback', that would have helped.